File: judo-sdk/src/main/java/com/judopay/card/TokenCardEntryFragment.java
package com.judopay.card;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Spinner;

import com.judopay.Judo;
import com.judopay.R;
import com.judopay.arch.ThemeUtil;
import com.judopay.model.Address;
import com.judopay.model.Card;
import com.judopay.model.CardToken;
import com.judopay.model.Country;
import com.judopay.validation.CountryAndPostcodeValidator;
import com.judopay.validation.SecurityCodeValidator;
import com.judopay.validation.Validation;
import com.judopay.validation.ValidationAutoAdvanceManager;
import com.judopay.validation.ValidationManager;
import com.judopay.validation.Validator;
import com.judopay.view.CardNumberEntryView;
import com.judopay.view.CountrySpinnerAdapter;
import com.judopay.view.ExpiryDateEntryView;
import com.judopay.view.PostcodeEntryView;
import com.judopay.view.SecurityCodeEntryView;
import com.judopay.view.SingleClickOnClickListener;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observables.ConnectableObservable;

public class TokenCardEntryFragment extends AbstractCardEntryFragment {
    private CardNumberEntryView cardNumberEntryView;
    private ExpiryDateEntryView expiryDateEntryView;
    private SecurityCodeEntryView securityCodeEntryView;
    private PostcodeEntryView postcodeEntryView;
    private Spinner countrySpinner;
    private View secureServerText;
    private View countryAndPostcodeContainer;
    private ValidationManager validationManager;

    private CompositeDisposable disposables = new CompositeDisposable();

    public static TokenCardEntryFragment newInstance(Judo judo, CardEntryListener listener) {
        TokenCardEntryFragment cardEntryFragment = new TokenCardEntryFragment();
        cardEntryFragment.setCardEntryListener(listener);

        Bundle arguments = new Bundle();
        arguments.putParcelable(Judo.JUDO_OPTIONS, judo);
        cardEntryFragment.setArguments(arguments);

        return cardEntryFragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_token_card_entry, container, false);

        this.submitButton = view.findViewById(R.id.button);
        this.cardNumberEntryView = view.findViewById(R.id.card_number_entry_view);
        this.expiryDateEntryView = view.findViewById(R.id.expiry_date_entry_view);
        this.securityCodeEntryView = view.findViewById(R.id.security_code_entry_view);
        this.postcodeEntryView = view.findViewById(R.id.postcode_entry_view);
        this.countryAndPostcodeContainer = view.findViewById(R.id.country_postcode_container);

        this.countrySpinner = view.findViewById(R.id.country_spinner);
        this.secureServerText = view.findViewById(R.id.secure_server_text);

        return view;
    }

    @Override
    protected void onInitialize(Bundle savedInstanceState, Judo judo) {
        CardToken cardToken = judo.getCardToken();

        if (cardToken == null) {
            throw new IllegalArgumentException("CardToken is required in Judo for TokenCardEntryFragment");
        }

        if (getActivity() != null) {
            boolean secureServerMessageShown = ThemeUtil.getBooleanAttr(getActivity(), R.attr.secureServerMessageShown);
            if (secureServerMessageShown) {
                secureServerText.setVisibility(View.VISIBLE);
            } else {
                secureServerText.setVisibility(View.GONE);
            }
        }

        initializeInputs(cardToken, judo);
        initializePayButton(judo);
        initializeCountry();

        initializeValidators(cardToken, judo);
    }

    @Override
    public void onDestroy() {
        disposables.clear();
        super.onDestroy();
    }

    private void initializeCountry() {
        countrySpinner.setAdapter(new CountrySpinnerAdapter(getActivity(), Country.avsCountries()));
    }

    private void initializeInputs(CardToken cardToken, Judo options) {
        cardNumberEntryView.setCardType(cardToken.getType(), false);
        securityCodeEntryView.setHelperText(R.string.please_reenter_the_card_security_code);
        securityCodeEntryView.setCardType(cardToken.getType(), false);
        securityCodeEntryView.requestFocus();

        expiryDateEntryView.setExpiryDate(options.getCardToken().getFormattedEndDate());
        expiryDateEntryView.setEnabled(false);
        cardNumberEntryView.setTokenCard(options.getCardToken());
    }

    private void initializeValidators(CardToken cardToken, final Judo judo) {
        List<Validator> validators = new ArrayList<>();
        List<Pair<Validator, View>> validatorViews = new ArrayList<>();

        SecurityCodeValidator securityCodeValidator = new SecurityCodeValidator(securityCodeEntryView.getEditText());
        securityCodeValidator.setCardType(cardToken.getType());
        disposables.add(securityCodeValidator.onValidate()
                .subscribe(validation -> {
                    if (judo.isAvsEnabled()) {
                        countryAndPostcodeContainer.setVisibility(validation.isValid() ? View.VISIBLE : View.GONE);
                    }
                }));

        validators.add(securityCodeValidator);
        validatorViews.add(new Pair<>(securityCodeValidator, securityCodeEntryView.getEditText()));

        validationManager = new ValidationManager(validators, this);

        if (judo.isAvsEnabled()) {
            initializeAvsValidators(validatorViews);
        }

        ValidationAutoAdvanceManager.bind(validationManager, validatorViews);
    }

    private void initializeAvsValidators(List<Pair<Validator, View>> validatorViews) {
        CountryAndPostcodeValidator countryAndPostcodeValidator = new CountryAndPostcodeValidator(countrySpinner, postcodeEntryView.getEditText());
        ConnectableObservable<Validation> observable = countryAndPostcodeValidator.onValidate();

        disposables.add(observable.subscribe(validation -> postcodeEntryView.setError(validation.getError(), validation.isShowError())));

        disposables.add(observable.subscribe(validation -> {
            String country = (String) countrySpinner.getSelectedItem();
            postcodeEntryView.setCountry(country);
        }));

        validationManager.addValidator(countryAndPostcodeValidator, observable);
        validatorViews.add(new Pair<>(countryAndPostcodeValidator, postcodeEntryView.getEditText()));

        observable.connect();
    }

    @Override
    public void onValidate(boolean valid) {
        submitButton.setVisibility(valid ? View.VISIBLE : View.GONE);
    }

    private void initializePayButton(final Judo judo) {
        submitButton.setOnClickListener(new SingleClickOnClickListener() {
            @Override
            public void doClick() {
                if (getActivity() != null) {
                    View view = getActivity().getCurrentFocus();
                    if (view != null) {
                        view.clearFocus();
                    }
                    hideKeyboard();
                    submitForm(judo);
                }
            }
        });
    }

    private void submitForm(Judo judo) {
        Card.Builder cardBuilder = new Card.Builder()
                .setCardNumber(cardNumberEntryView.getText())
                .setExpiryDate(expiryDateEntryView.getText())
                .setSecurityCode(securityCodeEntryView.getText());

        if (judo.isAvsEnabled()) {
            cardBuilder.setAddress(new Address.Builder()
                    .setPostCode(postcodeEntryView.getText())
                    .setCountryCode(Country.codeFromCountry((String) countrySpinner.getSelectedItem()))
                    .build());
        }

        if (cardEntryListener != null) {
            cardEntryListener.onSubmit(cardBuilder.build());
        }
    }
}

File: judo-sdk/src/main/java/com/judopay/card/TokenCardEntryFragment.kt
package com.judopay.card

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.os.Handler
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.judopay.Judo
import com.judopay.R
import com.judopay.arch.ThemeUtil
import com.judopay.model.Address
import com.judopay.model.Card
import com.judopay.model.CardToken
import com.judopay.model.Country
import com.judopay.validation.CountryAndPostcodeValidator
import com.judopay.validation.SecurityCodeValidator
import com.judopay.validation.ValidationAutoAdvanceManager
import com.judopay.validation.ValidationManager
import com.judopay.validation.Validator
import com.judopay.view.CountrySpinnerAdapter
import com.judopay.view.SingleClickOnClickListener
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.fragment_card_entry.button
import kotlinx.android.synthetic.main.fragment_card_entry.cancelButton
import kotlinx.android.synthetic.main.fragment_card_entry.secureServerText
import kotlinx.android.synthetic.main.view_card_entry_form.cardNumberEntryView
import kotlinx.android.synthetic.main.view_card_entry_form.countryPostcodeContainer
import kotlinx.android.synthetic.main.view_card_entry_form.countrySpinner
import kotlinx.android.synthetic.main.view_card_entry_form.expiryDateEntryView
import kotlinx.android.synthetic.main.view_card_entry_form.postcodeEntryView
import kotlinx.android.synthetic.main.view_card_entry_form.securityCodeEntryView
import java.util.*

class TokenCardEntryFragment : AbstractCardEntryFragment() {
    private var validationManager: ValidationManager? = null
    private val disposables = CompositeDisposable()
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? =
        inflater.inflate(R.layout.fragment_card_entry, container, false)


    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.setOnShowListener { dialogInterface: DialogInterface? ->
            val bottomSheet =
                dialog.findViewById<View>(R.id.design_bottom_sheet)
            BottomSheetBehavior.from(bottomSheet)
                .setState(BottomSheetBehavior.STATE_EXPANDED)
        }
        return dialog
    }

    override fun onInitialize(savedInstanceState: Bundle?, judo: Judo) {
        val cardToken = judo.cardToken
            ?: throw IllegalArgumentException("CardToken is required in Judo for TokenCardEntryFragment")
        if (activity != null) {
            val secureServerMessageShown =
                ThemeUtil.getBooleanAttr(activity, R.attr.secureServerMessageShown)
            if (secureServerMessageShown) {
                secureServerText.visibility = View.VISIBLE
            } else {
                secureServerText.visibility = View.GONE
            }
        }
        cancelButton.setOnClickListener { onCancel(requireDialog()) }
        initializeInputs(cardToken, judo)
        initializePayButton(judo)
        initializeCountry()
        initializeValidators(cardToken, judo)
    }

    override fun onDestroy() {
        disposables.clear()
        super.onDestroy()
    }

    private fun initializeCountry() {
        countrySpinner.adapter = CountrySpinnerAdapter(activity, Country.values())
    }

    private fun initializeInputs(cardToken: CardToken, options: Judo) {
        cardNumberEntryView.setCardType(cardToken.type, false)
        securityCodeEntryView.setCardType(cardToken.type, false)
        expiryDateEntryView.setExpiryDate(options.cardToken.formattedEndDate)
        expiryDateEntryView.isEnabled = false
        cardNumberEntryView.setTokenCard(options.cardToken)
    }

    private fun initializeValidators(cardToken: CardToken, judo: Judo) {
        val validators: MutableList<Validator> = ArrayList()
        val validatorViews: MutableList<Pair<Validator, View?>> = ArrayList()
        val securityCodeValidator = SecurityCodeValidator(securityCodeEntryView.editText)
        securityCodeValidator.setCardType(cardToken.type)
        disposables.add(securityCodeValidator.onValidate()
            .subscribe {
                if (judo.isAvsEnabled) {
                    countryPostcodeContainer.visibility =
                        if (it.isValid) View.VISIBLE else View.GONE
                }
            }
        )
        validators.add(securityCodeValidator)
        validatorViews.add(Pair(securityCodeValidator, securityCodeEntryView.editText))
        validationManager = ValidationManager(validators, this)
        if (judo.isAvsEnabled) {
            initializeAvsValidators(validatorViews)
        }
        ValidationAutoAdvanceManager.bind(validationManager, validatorViews)
    }

    private fun initializeAvsValidators(validatorViews: MutableList<Pair<Validator, View?>>) {
        val countryAndPostcodeValidator =
            CountryAndPostcodeValidator(countrySpinner, postcodeEntryView.editText)
        val observable = countryAndPostcodeValidator.onValidate()
        disposables.add(observable.subscribe {
            postcodeEntryView.setValidation(it)
        })
        disposables.add(observable.subscribe {
            val country = countrySpinner.selectedItem as Country
            postcodeEntryView.setCountry(country)
        })
        validationManager?.addValidator(countryAndPostcodeValidator, observable)
        validatorViews.add(Pair(countryAndPostcodeValidator, postcodeEntryView.editText))
        observable.connect()
    }

    override fun onValidate(valid: Boolean) {
        button.getButton().isEnabled = valid
    }

    private fun initializePayButton(judo: Judo) {
        button.getButton().setOnClickListener(object : SingleClickOnClickListener() {
            override fun doClick() {
                if (activity != null) {
                    val view = requireActivity().currentFocus
                    view?.clearFocus()
                    button.startLoading()
                    hideKeyboard()
                    submitForm(judo)
                }
            }
        })
    }

    private fun submitForm(judo: Judo) {
        val cardBuilder = Card.Builder()
            .setCardNumber(cardNumberEntryView.getText())
            .setExpiryDate(expiryDateEntryView.getText())
            .setSecurityCode(securityCodeEntryView.getText())
        if (judo.isAvsEnabled) {
            cardBuilder.setAddress(
                Address.Builder()
                    .setPostCode(postcodeEntryView.getText())
                    .setCountryCode((countrySpinner.selectedItem as Country).countryCode)
                    .build()
            )
        }
        cardEntryListener.onSubmit(cardBuilder.build())
    }

    companion object {
        fun newInstance(judo: Judo, listener: CardEntryListener): TokenCardEntryFragment {
            val cardEntryFragment = TokenCardEntryFragment()
            cardEntryFragment.cardEntryListener = listener
            val arguments = Bundle()
            arguments.putParcelable(Judo.JUDO_OPTIONS, judo)
            cardEntryFragment.arguments = arguments
            return cardEntryFragment
        }
    }
}